library/src/blas2/rocblas_syr.hpp

```diff
@@ -5,35 +5,79 @@
 #include "handle.h"
 #include "rocblas.h"
 
-template <typename T, typename U>
-__global__ void rocblas_syr_kernel(rocblas_fill uplo,
-                                   rocblas_int  n,
-                                   U            alpha_device_host,
-                                   const T* __restrict__ x,
-                                   rocblas_int incx,
-                                   T*          A,
-                                   rocblas_int lda)
+template <typename T, typename U, typename V, typename W>
+__global__ void rocblas_syr_kernel(rocblas_fill   uplo,
+                                   rocblas_int    n,
+                                   U              alpha_device_host,
+                                   rocblas_stride stride_alpha,
+                                   V              xa,
+                                   ptrdiff_t      shiftx,
+                                   rocblas_int    incx,
+                                   rocblas_stride stridex,
+                                   W              Aa,
+                                   ptrdiff_t      shiftA,
```

Comment:
T is not needed here, it can be removed with the use of the 'auto' keyword, partially used in this implementation, see x and A.

Reply:
I agree that `T` is not necessary. But rather than using `auto`, you can use `decltype` or traits to determine the type based on other, automatically deduced, template parameters. Such as:
```
using T = decltype(...); // Put an expression in (...) which evaluates to the type of T
T* A;
```
Or short of going complete `auto`, you can use `auto *` to indicate that it's a pointer.